File: packages/typescript/src/threatfeed/compare.ts
/**
 * High-level comparison API for threat feed matching.
 */

import type { SignatureResult } from '../types';
import type { ThreatFeedCache } from './cache';
import type { ThreatMatch } from './types';

/**
 * Compare a signature result against the threat feed cache.
 *
 * Extracts the primary signature (family 0) from the result and queries
 * the cache for similar known threat signatures.
 *
 * @param result Signature result from signText()
 * @param cache Pre-loaded threat feed cache
 * @param options Comparison options
 * @param options.threshold Minimum cosine similarity threshold (default: 0.85)
 * @param options.maxResults Maximum number of results to return (default: 10)
 * @returns Array of ThreatMatch objects sorted by cosine similarity descending
 *
 * @example
 * ```typescript
 * import { signText, SignatureVersion } from '@0din/prompt-toolkit';
 * import { ThreatFeedCache, compareToThreatfeed } from '@0din/prompt-toolkit/threatfeed';
 *
 * const result = await signText('suspicious prompt', provider, SignatureVersion.V1);
 * const cache = new ThreatFeedCache({ version: SignatureVersion.V1 });
 * cache.load();
 *
 * const matches = compareToThreatfeed(result, cache, { threshold: 0.85 });
 * for (const m of matches) {
 *   console.log(`Match: ${m.title} (similarity: ${m.cosineSimilarity.toFixed(3)})`);
 * }
 * ```
 */
export function compareToThreatfeed(
  result: SignatureResult,
  cache: ThreatFeedCache,
  options?: { threshold?: number; maxResults?: number },
): ThreatMatch[] {
  const primarySig = result.lsh.signatures[0].signature;
  return cache.query(primarySig, options);
}

File: packages/rust/src/threatfeed/compare.rs
//! High-level comparison API for threat feed matching.

use crate::error::Result;
use crate::types::SignatureResult;

use super::cache::ThreatFeedCache;
use super::types::ThreatMatch;

/// Compare a signature result against the threat feed cache.
///
/// Extracts the primary signature (family 0) from the result and queries
/// the cache for similar known threat signatures.
///
/// # Arguments
///
/// * `result` - Signature result from `sign_text()`
/// * `cache` - Pre-loaded threat feed cache
/// * `threshold` - Minimum cosine similarity threshold (default: 0.85)
/// * `max_results` - Maximum number of results to return (default: 10)
///
/// # Returns
///
/// Vector of matches sorted by cosine similarity descending.
///
/// # Example
///
/// ```rust,no_run
/// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
/// use odin_prompt_toolkit::{sign_text, SignatureVersion};
/// use odin_prompt_toolkit::threatfeed::{ThreatFeedCache, compare_to_threatfeed};
///
/// // Sign some text
/// let result = sign_text("suspicious prompt", SignatureVersion::V1, None, None).await?;
///
/// // Load the threat feed cache
/// let mut cache = ThreatFeedCache::new(SignatureVersion::V1, None, None);
/// cache.load()?;
///
/// // Compare against known threats
/// let matches = compare_to_threatfeed(&result, &cache, 0.85, 10)?;
/// for m in &matches {
///     println!("Match: {} (similarity: {:.3})", m.title, m.cosine_similarity);
/// }
/// # Ok(())
/// # }
/// ```
pub fn compare_to_threatfeed(
    result: &SignatureResult,
    cache: &ThreatFeedCache,
    threshold: f64,
    max_results: usize,
) -> Result<Vec<ThreatMatch>> {
    let primary_sig = &result.lsh.signatures[0].signature;
    Ok(cache.query(primary_sig, threshold, max_results))
}
